File: dotnet/src/MyDotey.CircularBuffer/TimeBucket.cs
using System;
using System.Threading;

namespace MyDotey.Util
{
    /**
     * @author koqizhao
     *
     * Jul 25, 2018
     */
    public abstract class TimeBucket
    {
        protected internal static long CurrentTimeMillis { get { return DateTime.Now.Ticks / TimeSpan.TicksPerMillisecond; } }

        private long _startTime;
        private long _ttl;

        protected TimeBucket(long startTime, long ttl)
        {
            _startTime = startTime;
            _ttl = ttl;
        }

        public long StartTime
        {
            get { return _startTime; }
        }

        public long Ttl
        {
            get { return _ttl; }
        }

        public virtual void Reset(long startTime)
        {
            Interlocked.Exchange(ref _startTime, startTime);
        }

        public bool IsStale
        {
            get { return CurrentTimeMillis - _startTime > _ttl; }
        }
    }
}

File: java/circular-buffer/src/main/java/org/mydotey/util/TimeBucket.java
package org.mydotey.util;

/**
 * @author koqizhao
 *
 * Jul 25, 2018
 */
public abstract class TimeBucket {

    private volatile long _startTime;
    private long _ttl;

    protected TimeBucket(long startTime, long ttl) {
        _startTime = startTime;
        _ttl = ttl;
    }

    public long getStartTime() {
        return _startTime;
    }

    public long getTtl() {
        return _ttl;
    }

    public void reset(long startTime) {
        _startTime = startTime;
    }

    public boolean isStale() {
        return System.currentTimeMillis() - _startTime > _ttl;
    }

}
